### PR TITLE
feat(metrics): segmentation confusion suite over filter buckets

### PR DESCRIPTION
**Pull request by amadeuszsz**
## Summary

Replaces the range-only segmentation suite with `Segmentation3DConfusionMatrixMetricSuite` over (filter, range) buckets.

- One fixed-size confusion matrix per (filter, range) bucket in a single stacked `(F+1, R+1, C, C)` sum state, so memory never grows with the dataset and DDP reduction stays exact.
- Consumes per-frame `seg_frames` entries (points plus ego pose and scene token), resolving per-point filter membership at update, with map coverage counters for scenes without a lanelet map.
- Adds the raw `ConfusionMatrix` component and per-class-group folding, replaces the flat point-tensor input path, and updates the distributed and segmentation tests to the new contract.

## Related Issues

Stacked on #102.

## Checklist

- [x] You've checked our [contribution overview](https://tier4.github.io/autoware-ml/contributing/contribution-overview/).
- [x] Your PR follows our [pull request guidelines](https://tier4.github.io/autoware-ml/contributing/pull-request-guidelines/).

## Additional Notes

Models emit `seg_frames` in a follow-up PR of this stack, until then the suite is exercised by its tests only.

## Additional Log and Media

`pytest autoware_ml/tests/{metrics,configs,datamodule}`: 177 passed, 1 skipped at this branch head.

### CODE REVIEW

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 1]
```diff
@@ -1,30 +1,33 @@
...
+from autoware_ml.metrics.class_groups import fold_confusion, resolve_class_groups
 from autoware_ml.metrics.segmentation3d.confusion import ConfusionState
 
 
-class Segmentation3DMetricSuite(MetricSuite[ConfusionState]):
-    """Confusion-matrix segmentation suite. Buckets points by range at update into
-    one stacked confusion state, and exposes a ``ConfusionState`` per bucket to the
-    injected metrics.
-    """
+class Segmentation3DConfusionMatrixMetricSuite(MetricSuite[ConfusionState]):
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py:26]: Right, the bundled configs still name the range-only suite. It is retained here alongside the new one and removed together with the last config reference in the config migration PR later in this stack, so every intermediate state instantiates.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 2]
```diff
@@ -1,30 +1,33 @@
...
+    """Confusion-matrix segmentation suite over (filter, range) buckets."""
 
     prefix = "seg3d"
-    _required_keys = ("seg_pred_labels", "seg_target_labels", "seg_coord")
+    _required_keys = ("seg_frames",)
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py:30]: That is the stacking order: the bundled configs keep instantiating the retained range-only suite (which consumes the current model outputs), the models start emitting seg_frames in a follow-up PR of this stack, and only the config migration PR after that points them at this suite. No shipped configuration reaches the seg_frames contract before its producers exist.

File: autoware_ml/transforms/boxes3d/annotations.py [hunk 3]
```diff
@@ -91,34 +108,44 @@ def resolve_detection_class(
...
+    pre-computed ``bbox_label_3d`` integer (which indexes the file's own,
+    possibly older, class table) is consulted *only* as a fallback when no raw
+    category name is recorded. This way changing the configured class set never
+    requires regenerating the info files: categories the current model does not
+    train are dropped by ``name_mapping``/``class_names``, and categories a file
+    predates are still picked up from their raw name.
+
+    Low-point boxes are not filtered here, that is the job of the point-count
+    filters (``min_num_points`` at train time, the metric suite at eval), which
+    subsume the lidar-point validity flag.
```
[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/transforms/boxes3d/annotations.py]: Agreed on the substance: with the validity flag gone, validation pipelines have no point filter, so zero-support boxes reach val/loss and with it checkpoint selection (the suites are fine, they apply min_num_points themselves). Both the flag removal and the training-only filter wiring are #96's lines, which sits under #100 while that one is still under review, so this waits for the rebase pass. The plan there is to give the validation pipelines the same range filter as training, with the far filter's upper bound set to .inf so the corners of the rectangular range are covered too.

[the remark this answers is not in the document]
Comment (amadeuszsz) [reply autoware_ml/transforms/boxes3d/annotations.py:126]: The validity flag stays out, the min-points filter does the job. Every detection and multi-task leaf now shares one point_filters list between the training pipeline and validation, so val/loss is computed on the same filtered targets the model trains on, while test keeps the unfiltered ground truth the occlusion suites need (val = test + point_filters through merge_lists). The far filters also lost their outer bound so the corners of the rectangular range are covered. The filter wiring lives in the multisweep PR, so that is where it landed: https://github.com/tier4/autoware-ml/pull/112/commits/4d7c105a2b820f173469819e3f2c337a439f682d.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 4]
```diff
@@ -1,22 +1,196 @@
...
+        self._group_lut, self._grouped_names = (
+            resolve_class_groups(self.class_names, class_groups) if class_groups else (None, None)
+        )
+
+        # Distinct evaluation filters used by the components. Bucket 0 is the
+        # unfiltered (whole-scene) slice, filter buckets follow in registration
+        # order.
+        self._register_component_filters()
+
+        # One matrix per (filter, range) bucket. Accumulation always uses the
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py:67]: This says the same as the comment at line 58 (fold is compute-time, synced state is always the trained classes). Keep one.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 5]
```diff
@@ -1,22 +1,196 @@
...
+                    continue
+                context = self._frame_filter_context(index, metric_filter, frame)
+                if context is None:
+                    continue  # scene without a lanelet map, excluded from this slice
+                keep = torch.from_numpy(
+                    np.asarray(metric_filter.keep(coord_np, context), dtype=bool)
+                ).to(predictions.device)
+                self._accumulate(index + 1, predictions[keep], targets[keep], coord[keep])
+
+    def _accumulate(
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py]: One-line docstrings for `_counts` and `_accumulate` please (what the `filter_bucket` index means and that ranges are applied on top of it).

**Comment by mojomex** [comment]
@amadeuszsz The PR's commit ee8626e seems to be buried under unrelated commits 3fee951, 3d0d10f - is this intended?

**Comment by amadeuszsz** [comment]
> @amadeuszsz The PR's commit [ee8626e](https://github.com/tier4/autoware-ml/commit/ee8626ea391189640b0e54af3cc5ec202b39105a) seems to be buried under unrelated commits [3fee951](https://github.com/tier4/autoware-ml/commit/3fee951b74fb395b9b303d5ef4b1b1a4d06bdead), [3d0d10f](https://github.com/tier4/autoware-ml/commit/3d0d10f8f07840d03d82e05c6cd12aa6eceed2ee) - is this intended?

Yes, this is just a sequence of stacked PRs.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 6]
```diff
@@ -1,22 +1,196 @@
...
+        class_names: tuple[str, ...] | None = None,
+        ranges: tuple[MetricRange, ...] = (),
+        class_groups: dict[str, list[str]] | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Register the stacked confusion state for every (filter, range) bucket.
+
+        Args:
+            components: Injected confusion-reading metrics, as in :class:`MetricSuite`.
+            num_classes: Number of trained classes, the matrix is ``C x C``.
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py]: Suggestion: rephrase to `Number ``C`` of trained classes, ...` so `num_classes` gets explicitly associated with `C`.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 7]
```diff
@@ -1,22 +1,196 @@
...
+            ignore_index: Target label excluded from the accumulation.
+            class_names: Ordered class names for metric keys, or ``None``.
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py]: `... of length ``C`` ...`

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 8]
```diff
@@ -1,22 +1,196 @@
...
+        counts = torch.bincount(indices, minlength=self.num_classes**2)
+        return counts.reshape(self.num_classes, self.num_classes).long()
+
+    def update(self, eval_out: dict[str, Any]) -> None:
+        """Fold one batch of per-frame point predictions into the matrices.
+
+        Args:
+            eval_out: Flat eval-output dict with the per-frame ``seg_frames``.
+        """
+        active = self._stage_filter_keys()
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py]: Please rename to `active_filter_keys` or similar - currently it reads like a boolean.

File: autoware_ml/metrics/segmentation3d/suite.py [hunk 9]
```diff
@@ -1,22 +1,196 @@
...
+                ).to(predictions.device)
+                self._accumulate(index + 1, predictions[keep], targets[keep], coord[keep])
+
+    def _accumulate(
+        self,
+        filter_bucket: int,
+        predictions: torch.Tensor,
+        targets: torch.Tensor,
+        coord: torch.Tensor,
+    ) -> None:
```
Comment (mojomex) [inline autoware_ml/metrics/segmentation3d/suite.py:147]: Short doc comment please explaining what filter_bucket is, and what will be updated (all ranges for the filter bucket).

File: autoware_ml/tests/metrics/test_confusion_matrix.py [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+from autoware_ml.metrics.confusion_report import confusion_cells
+from autoware_ml.metrics.detection3d.confusion_matrix import ConfusionMatrix as DetConfusion
+from autoware_ml.metrics.detection3d.matching import DetectionState
+from autoware_ml.metrics.detection3d.structures import Detection3DSample
+from autoware_ml.metrics.segmentation3d.confusion import ConfusionState
+from autoware_ml.metrics.segmentation3d.confusion_matrix import ConfusionMatrix as SegConfusion
+from autoware_ml.metrics.segmentation3d.suite import Segmentation3DConfusionMatrixMetricSuite
+
+
+def test_confusion_cells_flattens_true_by_pred() -> None:
```
Comment (mojomex) [inline autoware_ml/tests/metrics/test_confusion_matrix.py]: This name is hard to understand - is it referring to the metric token being flattened to `confusion_<true>__<pred>`?

File: autoware_ml/tests/metrics/test_confusion_matrix.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+                    "target": torch.tensor([1, 0, 3, 2]),  # all intra-group swaps
+                    "coord": torch.zeros((4, 3)),
+                }
+            ]
+        }
+    )
+    report = suite.result(EvalStage.TEST)
+    # Both vehicle points land in the grouped_vehicle diagonal, none leak to flat.
+    assert report["confusion_grouped_vehicle__grouped_vehicle"] == 2.0
+    assert report["confusion_grouped_vehicle__grouped_flat"] == 0.0
```
Comment (mojomex) [inline autoware_ml/tests/metrics/test_confusion_matrix.py:92]: For completeness sake, might also want to check that flat__vehicle is 0.

**Review by mojomex — COMMENTED** [review]
LGTM! All comments are nits.

File: autoware_ml/metrics/segmentation3d/suite.py (hunk 4 above)
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py:67]: Dropped this one, the comment on the state keeps the reason. Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/d6c08b1610afc82dfb57b9ee3719b947163f0e23.

File: autoware_ml/metrics/segmentation3d/suite.py (hunk 6 above)
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py]: Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/d6c08b1610afc82dfb57b9ee3719b947163f0e23.

File: autoware_ml/metrics/segmentation3d/suite.py (hunk 7 above)
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py]: Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/d6c08b1610afc82dfb57b9ee3719b947163f0e23.

File: autoware_ml/metrics/segmentation3d/suite.py (hunk 8 above)
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py]: Renamed to `active_filter_keys`. Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/9feee7258d58e48050f907ff9846a8342bca7180.

File: autoware_ml/metrics/segmentation3d/suite.py (hunk 9 above)
Comment (amadeuszsz) [reply autoware_ml/metrics/segmentation3d/suite.py:147]: Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/9feee7258d58e48050f907ff9846a8342bca7180.

File: autoware_ml/tests/metrics/test_confusion_matrix.py (hunk 10 above)
Comment (amadeuszsz) [reply autoware_ml/tests/metrics/test_confusion_matrix.py]: Yes, the `confusion_<true>__<pred>` key layout. Renamed so the name says it. Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/8f518081e8100fe7b6f6735897d869ec16461a45.

File: autoware_ml/tests/metrics/test_confusion_matrix.py (hunk 11 above)
Comment (amadeuszsz) [reply autoware_ml/tests/metrics/test_confusion_matrix.py:92]: Addressed in https://github.com/tier4/autoware-ml/pull/103/commits/8f518081e8100fe7b6f6735897d869ec16461a45.

**Review by mojomex — DISMISSED** [review]

**Review by mojomex — DISMISSED** [review]
